Convert ActionCable library from CoffeeScript to JavaScript

varyonic opened the pull request:
In response to #33331

varyonic committed:
Rename actioncable test coffee files to js.

varyonic committed:
Rename actioncable coffee files to js.
Move javascript out of app/assets

varyonic committed:
Rename actioncable coffee files to js.

rafaelfranca commented:
If we would do this, and only if, it would be to use ES6 like Active Storage do. Changing CoffeScript with ES5 is not buying us anything,

javan commented:
I agree with @rafaelfranca and suggest using the exact rollup+babel+eslint rig from Active Storage: https://github.com/rails/rails/blob/e13e16f5adf3221f4a95e7d105efb2d5e56c0b53/activestorage/package.json#L35-L38

varyonic committed:
Machine translate actioncable coffeescript to ES6.

varyonic committed:
Machine translate actioncable coffeescript to ES6.
Use ES6 modules.

varyonic committed:
Machine translate actioncable test coffeescript to ES6.

varyonic committed:
Machine translate actioncable test coffeescript to ES6.

rafaelfranca commented:
Great! @varyonic are you going to work on it?

varyonic committed:
Build ES6 package using Rollup.

varyonic committed:
Build ES6 package using Rollup.

varyonic committed:
Use ES6 modules and remove semi-colons.

varyonic committed:
Move javascript out of app/assets

varyonic committed:
Build output actioncable.js

varyonic committed:
Build output actioncable.js transpiled correctly to ES5.

varyonic commented:
@javan Does Blade need to be replaced with Karma or similar here? If so any examples I should look at for guidance?

varyonic committed:
npm install karma qunitjs mock-socket --save-dev
karma.conf for Rollup/RequireJS plus boilerplate test page to load and run unit tests.

javan commented on actioncable/app/javascript/actioncable/helpers.js:
```diff
@@ -0,0 +1,16 @@
+  if (this.debugging) {
+    messages.push(Date.now())
+    this.logger = window.console
+    return this.logger.log("[ActionCable]", ...Array.from(messages))
```
Don't need `Array.from()` here: `this.logger.log("[ActionCable]", ...messages)`

javan commented on actioncable/app/javascript/actioncable/connection_monitor.js:
```diff
@@ -0,0 +1,129 @@
+
+let now = undefined
+let secondsSince = undefined
+let clamp = undefined
```
No need to start these out `undefined` and then set them later (immediately, basically).

javan commented on actioncable/app/javascript/actioncable/helpers.js:
```diff
@@ -0,0 +1,16 @@
+
+export function stopDebugging() {
+  return this.debugging = null
+}
```
The `debugging` flag should only be set on the top level object, but these will set it on whatever happens to be `this`. I'd avoid using `this` in helpers, generally.

javan commented on actioncable/app/javascript/actioncable/index.js:
```diff
@@ -0,0 +1,47 @@
+
+  createConsumer(url) {
+    if (url == null) { let left
+    url = (left = this.getConfig("url")) != null ? left : this.INTERNAL.default_mount_path }
```
Is `INTERNAL` defined anywhere? Will need some kind of setup to replace: https://github.com/rails/rails/blob/45903be1344992b1f0156cd88c351067bdc93726/actioncable/app/assets/javascripts/action_cable.coffee.erb#L5-L6

javan commented on actioncable/app/javascript/actioncable/subscriptions.js:
```diff
@@ -0,0 +1,94 @@
+  reject(identifier) {
+    return (() => {
+      const result = []
+      for (let subscription of Array.from(this.findAll(identifier))) {
```
I *think* there are browsers that support `WebSocket`, but not `Array.from()` so we should avoid using it if possible.

javan reviewed (changes requested):
Great start!

Generally, the code feels like a direct translation of the CoffeeScript and not necessarily how you'd write it from scratch in ES6.

> Does Blade need to be replaced with Karma or similar here?

Yes, please. :) We should remove Blade / Sprockets dependencies entirely. I'll try to find you a good example to reference.

varyonic committed:
javan requested changes

varyonic commented on actioncable/app/javascript/actioncable/connection.js:
```diff
@@ -0,0 +1,161 @@
+//  ,unsupportedProtocol = protocols[adjustedLength - 1]
+
+export class Connection {
+  static initClass() {
```
Is this valid ES6? The code feels like a direct translation because it is. I was hoping to have working tests before refactoring.

javan replied:
Not yet… You can use the [transform-class-properties](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/) Babel plugin. Or, just assign those properties after the class definition:
```js
export class Connection {  
  // …
}

Connection.reopenDelay = 500
```

varyonic commented:
I added and configured karma-qunit but the tests are erroring "Uncaught Error: pushFailure() assertion outside test context" I think that's something to do with mock-socket not being setup correctly but I don't understand how to troubleshoot it.

javan commented:
For the Karma setup, I think it'd be more straightforward to compile a single .js file *first* and then start the runner. That way you don't need to muck around with requirejs and stash a list of a files globally. See https://github.com/josh/selector-observer for a good example of this.

oniofchaos commented:
Two thoughts:

1) Is this something that could be done on a more incremental level so it's not one big PR that sits around for ages?

2) Does it need to be tagged with the Rails 6 milestone?

varyonic commented:
See #34177
